Spark3 structured streaming enable updates

karim-ramadan opened the pull request.

aokolnychyi committed:
Spark 3.4: Initial support

hililiwei committed:
Flink: Remove Flink 1.14

hililiwei committed:
Flink: Make flink 1.17 work

Commit:
spark3: Fixed unit tests on Windows OS

hililiwei committed:
Flink: Move flink/v1.16 to flink/v1.17

hililiwei committed:
Flink: Copy flink/1.16 files from flink/1.17

Commit:
Spark3: Added option streaming-overwrite-snapshots-read-mode with 3 modes (BREAK, SKIP, ADDED_FILES_ONLY) to model the behaviour a streaming reader should have when encountering an overwrite snapshot.

Commit:
Spark3: Added unit tests for conf "streaming-overwrite-snapshots-read-mode"

tmnd1991 commented on core/src/main/java/org/apache/iceberg/StreamingOverwriteSnapshotReadMode.java:
```diff
@@ -0,0 +1,39 @@
+  ADDED_FILES_ONLY;
+
+  public static StreamingOverwriteSnapshotReadMode fromName(String readMode) {
+    Preconditions.checkArgument(readMode != null, "Invalid streaming overwrite snapshot read mode");
```
```suggestion
    Preconditions.checkArgument(readMode != null, "`null` is not a valid streaming overwrite snapshot read mode");
```

tmnd1991 commented on hive-metastore/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java:
```diff
@@ -292,7 +293,11 @@ public void testCreateTableTxnWithGlobalTableLocation() {
 
     Transaction txn =
         catalog.newCreateTableTransaction(
-            TABLE_IDENTIFIER, SCHEMA, SPEC, "file:///" + tableLocation, Maps.newHashMap());
```
are you sure about indentation? Probably iceberg has some maven/gradle goal to reformat as expected

karim-ramadan replied:
I'm sure I have configured google java format

tmnd1991 commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkReadConf.java:
```diff
@@ -132,14 +134,34 @@ public boolean streamingSkipDeleteSnapshots() {
   }
 
-  public boolean streamingSkipOverwriteSnapshots() {
+  private boolean streamingSkipOverwriteSnapshots() {
```
why did you change the visibility of this method? I think it's a blocker for a revision to change the public interface

karim-ramadan replied:
I intend to deprecate the old conf and just keep the new one, anyway this was the only usage of the method

tmnd1991 commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogTestBase.java:
```diff
@@ -49,8 +49,6 @@ public static Object[][] parameters() {
     };
   }
 
-  @Rule public TemporaryFolder temp = new TemporaryFolder();
```
please leave this change to another PR if possible

tmnd1991 commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/TestSparkReadConf.java:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  @Test
+  public void testStreamingOverwriteSnapshotReadModeDefaultShouldWinOverOlderConf() {
```
keeping two configurations, makes this interaction a bit weird, but I have no clear idea about it. Probably I would remove `SparkReadOptions.STREAMING_SKIP_OVERWRITE_SNAPSHOTS`

tmnd1991 replied:
also I would expect that if skip is true, the read mode is ignored

karim-ramadan replied:
Removing the old conf would be best I don't know if can be done

tmnd1991 replied:
when opening the PR, mention it :)

tmnd1991 commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/TestSparkReadConf.java:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  @Test
+  public void testStreamingOverwriteSnapshotReadModeDefaultShouldParseRegardlessOfCasing() {
```
same here

tmnd1991 commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/actions/TestCreateActions.java:
```diff
@@ -19,7 +19,6 @@
 package org.apache.iceberg.spark.actions;
```
Consider moving this changes to a separated PR

karim-ramadan committed:
Spark3: Applied reformat

dramaticlly committed:
Core: Refactor PartitionsTable planning (#7190)

deepyaman committed:
Python: Use Prettier to format TOML (#7248)

* Python: Use Prettier to format TOML and more files

* Python: Make Prettier skip the mkdocs admonoitions

dependabot[bot] committed:
Build: Bump coverage from 7.2.2 to 7.2.3 in /python (#7308)

dependabot[bot] committed:
Build: Bump pytest from 7.2.2 to 7.3.0 in /python (#7306)

dependabot[bot] committed:
Build: Bump pre-commit from 3.2.1 to 3.2.2 in /python (#7307)

ChristinaTech committed:
AWS: Check commit status after failed commit if AWS client performed retries (#7198)

ajantha-bhat committed:
Core: Fix errorprone warning (#7286)

snazy committed:
Build: Bump Nessie to 0.56.0 (#7283)

dependabot[bot] committed:
Build: Bump actions/stale from 7.0.0 to 8.0.0 (#7200)

Bumps [actions/stale](https://github.com/actions/stale) from 7.0.0 to 8.0.0.
- [Release notes](https://github.com/actions/stale/releases)
- [Changelog](https://github.com/actions/stale/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/stale/compare/v7.0.0...v8.0.0)

---
updated-dependencies:
- dependency-name: actions/stale
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Build: Bump org.apache.hadoop:hadoop-client from 3.3.4 to 3.3.5 (#7201)

Bumps org.apache.hadoop:hadoop-client from 3.3.4 to 3.3.5.

---
updated-dependencies:
- dependency-name: org.apache.hadoop:hadoop-client
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

bryanck committed:
Spark 3.1, 3.2: Broadcast Table instead of FileIO in rewrite manifests (#7263) (#7296)

This change backports PR #7263 to Spark 3.1 and 3.2

doki23 committed:
Build: `iceberg-delta-lake` Spark version to 3.3.2 (#7199)

ajantha-bhat committed:
Nessie: Use latest hash for catalog APIs (#6789)

* Nessie: Handle refresh for catalog APIs that doesn't use table operations

* Add commit testcase

* Another test case

* Address comments

* Avoid hash roundtrip

* Address new comments

* refactor

yabola committed:
Arrow: Support vectorized read of INT96 timestamp in imported data (#6962)

hililiwei committed:
Flink: Expose write-parallelism in SQL Hints (#7039)

Co-authored-by: chidayong <247070443@qq.com>

ajantha-bhat committed:
Nessie: Fix testcase failures (#7320)

stevenzwu committed:
Flink: move the classes from flink.sink.shuffle.statistics pkg to one level up as flink.sink.shuffle pkg (#7322)

flyrain committed:
Spark 3.3: Add doc for the changelog view procedure. (#7147)

hililiwei committed:
Flink 1.15: Port Expose write-parallelism in SQL Hints to 1.15

hililiwei committed:
Flink 1.17: Port Expose write-parallelism in SQL Hints to 1.17

snazy committed:
Build: Bump Nessie from 0.56.0 to 0.57.0 (#7323)

danielcweeks committed:
Update issue template for 1.2.1 release (#7331)

zhongyujiang committed:
Core: Fix SnapshotProducer#targetBranch exception message (#7315)

snazy committed:
Build: Bump Gradle from 8.0.2 to 8.1 (#7333)

ajantha-bhat committed:
Build: Fix flaky checkstyle issue (#7321)

Commit:
Infra: Update vote mail sample in source-release.sh (#7330)

nastra committed:
Core: Add missing metrics reporters when creating BaseTable (#7341)

aokolnychyi committed:
Core, Spark 3.3: Add FileRewriter API (#7175)

gustavoatt committed:
Spark: Accept an `output-spec-id` to write to a desired partition spec (#7120)

pavibhai committed:
ORC, Spark 3.3: Support selected vector with ORC row and batch readers (#7197)

chenjunjiedada committed:
Flink: use correct scan mode when in TABLE_SCAN_THEN_INCREMENTAL mode (#7338)

dependabot[bot] committed:
Build: Bump rich from 13.3.3 to 13.3.4 in /python (#7354)

Bumps [rich](https://github.com/Textualize/rich) from 13.3.3 to 13.3.4.
- [Release notes](https://github.com/Textualize/rich/releases)
- [Changelog](https://github.com/Textualize/rich/blob/master/CHANGELOG.md)
- [Commits](https://github.com/Textualize/rich/compare/v13.3.3...v13.3.4)

---
updated-dependencies:
- dependency-name: rich
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Build: Bump pytest from 7.3.0 to 7.3.1 in /python (#7355)

Bumps [pytest](https://github.com/pytest-dev/pytest) from 7.3.0 to 7.3.1.
- [Release notes](https://github.com/pytest-dev/pytest/releases)
- [Changelog](https://github.com/pytest-dev/pytest/blob/main/CHANGELOG.rst)
- [Commits](https://github.com/pytest-dev/pytest/compare/7.3.0...7.3.1)

---
updated-dependencies:
- dependency-name: pytest
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Build: Bump moto from 4.1.6 to 4.1.7 in /python (#7356)

Bumps [moto](https://github.com/getmoto/moto) from 4.1.6 to 4.1.7.
- [Release notes](https://github.com/getmoto/moto/releases)
- [Changelog](https://github.com/getmoto/moto/blob/master/CHANGELOG.md)
- [Commits](https://github.com/getmoto/moto/compare/4.1.6...4.1.7)

---
updated-dependencies:
- dependency-name: moto
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

ericlgoodman committed:
AWS: Fix Glue catalog integration with SparkSessionCatalog (#7277)

Fokko committed:
Build: Bump Airlift from 0.21 to 0.24 (#7347)

They removed unnecessary links to the Hadoop codebase: https://github.com/airlift/aircompressor/commits/master

preaudc committed:
Docs: clarify Hive on Tez configuration (#7282)

gustavoatt committed:
Spark: Simplify checks of output-spec-id in SparkWriteConf (#7348)

dramaticlly committed:
Core: Fix variable name in SetDefaultPartitionSpec (#7350)

Co-authored-by: Steve Zhang <hongyue_zhang@apple.com>

singhpk234 committed:
Core, Spark: Make ObjectStoreLocationProvider serializable (#7353)

szehon-ho committed:
Core: Parameterize RewriteDataFile's CommitService (#7343)

aokolnychyi committed:
Spark 3.4: Move 3.3 classes to 3.4 folder

aokolnychyi committed:
Spark 3.3: Copy 3.3 classes back

amogh-jahagirdar committed:
Core: Fix flaky TestParallelIterable test (#7372)

Fokko committed:
Flink: Apply row level filtering (#7109)

* Flink: Apply row level filtering

* Fix the tests

* Add test for case-sensitive

* Reduce duplication using a private method

amogh-jahagirdar committed:
Spark: Surface better error message during streaming planning when checkpoint snapshot not found (#6480)

chenjunjiedada committed:
Flink: backport #7338 to 1.16 and 1.15 (#7373)

karuppayya committed:
Spark 3.3: Honor Spark case sensitivity in ALTER TABLE ... WRITE ORDERED (#7324)

aokolnychyi committed:
Spark 3.3: Surface better error message during streaming planning when checkpoint snapshot not found (#7381)

This change backports PR #6480 to 3.3.

Co-authored-by: Amogh Jahagirdar <jahamogh@amazon.com>

Fokko committed:
Spark 2.4: Remove module (#7385)

Fokko committed:
Build: Bump Hive to 2.3.9 (#7374)

Some small bug fixes, but also removes the `commons-httpclient` dependency, which is nice.

https://issues.apache.org/jira/secure/ReleaseNote.jspa?version=12350009&styleName=Text&projectId=12310843

nastra committed:
Core: Introduce CompositeMetricsReporter (#7337)

This also switches from a collection of reporters to the
`CompositeMetricsReporter` in `TableScanContext` to fix an issue where
duplicate reporters were not handled.

linyanghao committed:
Flink: Use starting sequence number by default when rewriting data files (#7218)

Co-authored-by: linyanghao <linyanghao.data@bytedance.com>

yegangy0718 committed:
Backport to Flink 1.15: Implement DataStatisticsOperator operator to collect and send traffic distribution for guiding smart shuffling (#7400)

Co-authored-by: gang_ye <gang_ye@apple.com>

Fokko committed:
Flink: Backport row filter into 1.15 and 1.16 (#7397)

singhpk234 committed:
Spark 3.3: support rate limit in Spark Streaming (#4479)

danielcweeks committed:
Spec: Add S3 configuration to REST table load (#7401)

JonasJ-ap committed:
Doc: Add a page explaining migration from other table formats to iceberg (#6600)

JonasJ-ap committed:
Doc: Fix typo in hive_migration.md (#7407)

singhpk234 committed:
Spark: Fix failing unit test (#7414)

Co-authored-by: Prashant Singh <psinghvk@amazon.com>

linyanghao committed:
Backport #7218 (#7404)

Co-authored-by: linyanghao <linyanghao.data@bytedance.com>

dependabot[bot] committed:
Build: Bump zstandard from 0.20.0 to 0.21.0 in /python (#7411)

Bumps [zstandard](https://github.com/indygreg/python-zstandard) from 0.20.0 to 0.21.0.
- [Release notes](https://github.com/indygreg/python-zstandard/releases)
- [Changelog](https://github.com/indygreg/python-zstandard/blob/main/docs/news.rst)
- [Commits](https://github.com/indygreg/python-zstandard/compare/0.20.0...0.21.0)

---
updated-dependencies:
- dependency-name: zstandard
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

edgarRd committed:
MR: Skip filter translation if there are no leaves (#7123)

LuigiCerone committed:
Python: Add mkdocstrings (#7108)

* Add mkdocstrings

* Add mkdocs theme

* Add theme to mkdocs and fix pages gen

* Fix missing license

* Fix prettier

munendrasn committed:
AWS: Fix default warehouse path in Dynamodb catalog (#7358)

stevenzwu committed:
Flink: sync 1.16 with 1.17 for backports missed or not ported identically (#7403)

stevenzwu committed:
Flink: sync 1.15 with 1.17 for missed backports previously (#7402)

rdblue committed:
Views: Clean up and clarify the view spec (#7416)

amogh-jahagirdar committed:
Docs: Separate page for Branching and Tagging (#6723)

rdblue committed:
Views: Fix SQL view representation field name (#7417)

pvary committed:
Hive: Use EnvironmentContext instead of Hive Locks to provide transactional commits after HIVE-26882 (#6570)

deepyaman committed:
Python: Provide version ranges for each dependency (#7376)

* Python: Provide version ranges for each dependency

* Python: "Fix" the range specification for `duckdb`

* Python: Add support for Pandas 2 and PyArrow 7.0.0

* Python: Remove extraneous versioning justification

* Python: Define explicit upper bound for Dependabot

* Python: Update PyArrow lower bound from 7.0 to 9.0

Co-authored-by: Fokko Driesprong <fokko@apache.org>

---------

Co-authored-by: Fokko Driesprong <fokko@apache.org>

amogh-jahagirdar committed:
Spark: Backport #6480 to Spark 3.2 and Spark 3.1 (#7425)

amogh-jahagirdar committed:
API, Core: Move schemaID from ViewRepresentation to ViewVersion and make it required (#7421)

aokolnychyi committed:
Spark 3.4: Relax constraints in SparkPartitioningAwareScan (#7423)

nastra committed:
Core: Extract REST metrics reporter into its own class (#7339)

aokolnychyi committed:
Spark 3.4: Add tests for SPJ when partition keys mismatch (#7424)

karuppayya committed:
Spark 3.4: Cherry pick case sensitivity backport (#7324)

singhpk234 committed:
Build: Run Iceberg with JDK 17 (#7391)

kevinjqliu committed:
Python: Use fsspec implementation of LocalFileSystem (#7432)

instead of pyarrow implementation of LocalFileSystem

nastra committed:
Views: Move 'operation' check to ViewVersion (#7428)

911432 committed:
Docs: Add Java 17 to README.md (#7434)

* Updated README.md

#7391 
It's just a simple change in my thinking.

* Reduce writing

I like that it's shorter.

Co-authored-by: Fokko Driesprong <fokko@apache.org>

---------

Co-authored-by: Fokko Driesprong <fokko@apache.org>

frankliee committed:
Hive: Clean up expired metastore clients (#7310)

nastra committed:
Core: Make TableScanContext immutable (#5985)

Co-authored-by: Liwei Li <hililiwei@gmail.com>

nastra committed:
Core: Move table-creation-without-namespace-test to CatalogTests (#7349)

singhpk234 committed:
Spark: Refactor SparkReadConf to use primitive type (#7429)

Co-authored-by: Prashant Singh <psinghvk@amazon.com>

aokolnychyi committed:
Spark 3.4: Remove deprecated classes (#7448)

nastra committed:
Arrow: Convert dict encoded vectors to their expected Arrow vector types (#3024)

vishnukumarsinha committed:
Spark: Fixed Typo in Spark Read Option Vectorization Javadoc (#7439)

aokolnychyi committed:
Spark 3.4: Remove no longer needed write extensions (#7443)

JonasJ-ap committed:
Delta: Add version and timestamp tags for each Delta Lake transaction (#7450)

JonasJ-ap committed:
Delta: Fix snapshotDataFilesCount and use Immutable for result (#7454)

DarthData410 committed:
Python: fix `mkdocs` warnings from `conversions.py` (#7462)

* updates to remove mkdocs warnings from_bytes()

* updated comments for mkdocs WARNINGS to be removed

singhpk234 committed:
Spark 3.4: Support rate limit in Spark Streaming (#7422)

DarthData410 committed:
Python: fix mkdocs warnings from visitors.py (#7464)

* updates to remove mkdocs warnings from_bytes()

* updated comments for mkdocs WARNINGS to be removed

* removal of warnings from mkdocs serve --strict

DarthData410 committed:
Python: fix mkdocs warnings from sorting.py (#7467)

* updates to remove mkdocs warnings from_bytes()

* updated comments for mkdocs WARNINGS to be removed

* removal of warnings from mkdocs serve --strict

* changes to SortOrder init() added order_id arg

* reverted...

* updated for keyword args vs args order_id

* removed blank line 123 fom sorting.py

* finalizing comments

* Remove spaces

---------

Co-authored-by: Fokko Driesprong <fokko@apache.org>

singhpk234 committed:
Spark: Fix failing Spark streaming rate limit unit test (#7470)

Co-authored-by: Prashant Singh <psinghvk@amazon.com>

dependabot[bot] committed:
Build: Bump ray from 2.3.1 to 2.4.0 in /python (#7479)

dependabot[bot] committed:
Build: Bump mkdocs-gen-files from 0.4.0 to 0.5.0 in /python (#7478)

dependabot[bot] committed:
Build: Bump mkdocstrings-python from 0.8.3 to 0.9.0 in /python (#7475)

dependabot[bot] committed:
Build: Bump pyyaml from 5.4.1 to 6.0 in /python (#7477)

Bumps [pyyaml](https://github.com/yaml/pyyaml) from 5.4.1 to 6.0.
- [Release notes](https://github.com/yaml/pyyaml/releases)
- [Changelog](https://github.com/yaml/pyyaml/blob/master/CHANGES)
- [Commits](https://github.com/yaml/pyyaml/compare/5.4.1...6.0)

---
updated-dependencies:
- dependency-name: pyyaml
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Build: Bump mkdocs-material from 9.1.5 to 9.1.8 in /python (#7476)

Bumps [mkdocs-material](https://github.com/squidfunk/mkdocs-material) from 9.1.5 to 9.1.8.
- [Release notes](https://github.com/squidfunk/mkdocs-material/releases)
- [Changelog](https://github.com/squidfunk/mkdocs-material/blob/master/CHANGELOG)
- [Commits](https://github.com/squidfunk/mkdocs-material/compare/9.1.5...9.1.8)

---
updated-dependencies:
- dependency-name: mkdocs-material
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

DarthData410 committed:
Python: Fix warnings in documentation (#7472)

* updates to remove mkdocs warnings from_bytes()

* updated comments for mkdocs WARNINGS to be removed

* removal of warnings from mkdocs serve --strict

* changes to SortOrder init() added order_id arg

* reverted...

* updated for keyword args vs args order_id

* removed blank line 123 fom sorting.py

* finalizing comments

* Remove spaces

* fixed all mkdocs serve --strict warnings

* fixed trailing whitespace

* updates to comments for mkdocs serve --strict mode

* updates to io/__init__.py for mkdocs warns removal

* updates to remove mkdocs warns from types.py

* updates for schema.py to remove mkdocs warns

* updates to remove mkdoc warns from literals.py

* final mkdocs fixed transforms.py

* request PR changes from code review

* reuest format changes for comments

---------

Co-authored-by: Fokko Driesprong <fokko@apache.org>

aokolnychyi committed:
Spark 3.4: Switch to built-in DELETE implementation (#7453)

DarthData410 committed:
Python: Fix code highlighting in docs (#7484)

* updates to remove mkdocs warnings from_bytes()

* updated comments for mkdocs WARNINGS to be removed

* removal of warnings from mkdocs serve --strict

* changes to SortOrder init() added order_id arg

* reverted...

* updated for keyword args vs args order_id

* removed blank line 123 fom sorting.py

* finalizing comments

* Remove spaces

* fixed all mkdocs serve --strict warnings

* fixed trailing whitespace

* updates to comments for mkdocs serve --strict mode

* updates to io/__init__.py for mkdocs warns removal

* updates to remove mkdocs warns from types.py

* updates for schema.py to remove mkdocs warns

* updates to remove mkdoc warns from literals.py

* final mkdocs fixed transforms.py

* request PR changes from code review

* reuest format changes for comments

* fix pymdownx markdown_ext for code hihglights

---------

Co-authored-by: Fokko Driesprong <fokko@apache.org>

liuxiaocs7 committed:
Spark: Remove Usage of deprecated AssertHelpers in spark-sql (#7486)

nastra committed:
Spark 3.3: Remove deprecated FileScanTaskSetManager (#7489)

Fokko committed:
Python: Update pre-commit to the latest version (#7436)

Removed some ignores and removed some list materialization

szehon-ho committed:
Hive: Support connecting to multiple HMS-Catalog on same HMS URL (#7441)

nastra committed:
Spark 3.4: Add read and write support for UUIDs (#7399)

JonasJ-ap committed:
Python: Infer Iceberg schema from the Parquet file (#6997)

liuxiaocs7 committed:
Hive: Remove deprecated AssertHelpers (#7482)

liuxiaocs7 committed:
Flink: Remove deprecated AssertHelpers (#7481)

liuxiaocs7 committed:
Spark: Remove deprecated AssertHelpers usage (#7483)

karim-ramadan committed:
core: Applied spotless format

karim-ramadan committed:
spark3: Applied spotless format

karim-ramadan committed:
spark3: Fixed unit test for STREAMING_OVERWRITE_SNAPSHOTS_READ_MODE config
